packages/opentelemetry-core/src/default_span.ts

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Span,
+  SpanContext,
+  Status,
+  Attributes,
+  SpanKind,
+} from '@opentelemetry/types';
+
+/** This is a no-op implementation of Span. */
+export class DefaultSpan implements Span {
+  protected name = 'DefaultSpan'; // default
+  protected spanId = '';
+  protected parentSpanId = '';
+  protected kind = SpanKind.INTERNAL; // default
+  protected spanContext: SpanContext | null = null;
+
+  constructor() {
+    // TODO: https://github.com/open-telemetry/opentelemetry-js/pull/35
+    // this.spanId = randomSpanId();
+  }
+
+  context(): SpanContext {
+    return this._context();
+  }
+
+  setAttribute(key: string, value: unknown): this {
```

Comment:
`value` should be optional. I missed that on the interface but we have use cases where we don't explicitly pass a value.

Reply:
Ok. For the sake of record, Is it possible to share the use cases here?

Reply:
Also - what should be stored in the underlying data structure in that case @rochdev?
Undefined, null, ...?

Reply:
> Ok. For the sake of record, Is it possible to share the use cases here?

We use tags in OpenTracing to enable certain functionalities. But having to pass `true` or `1` explicitly is not needed for these tags since it's implicit.

For example:

```js
span.setAttribute('feature.enabled')
```

> Also - what should be stored in the underlying data structure in that case @rochdev?

This depends on the implementation. In our case, for the attributes where we support this we would set it to `true` and for other tags it would simply use the value as is, effectively removing the attribute by assigning `undefined`.

Reply:
There is also the fact that `undefined` is a valid value for `unknown` so it would be possible to explicitly pass `undefined` anyway. Making the value optional just makes it easier to pass `undefined`.

Reply:
I would suggest to open a separate issue to address this later. May be when we do the first revision.

Reply:
if we allow no values, may we need to deal with them in exporters, I'm not sure every backend accepts no value. It also means we may need to set a "no value" when we report the span like an empty string.

Reply:
Some backends will definitely not support no value. In that case, no value can either be converted to a string `"undefined"` or simply ignored.

Reply:
I think OpenTracing had the right idea. The spec says that the value can be anything, and it's up to the implementor to deal with whatever value is received.

Reply:
A more concrete example from OpenTracing was for example `inject` and `extract` which accept either a span or span context, but `_inject` and `_extract` are guaranteed to receive a span context.